File: backend/app/schemas/insurance.py
from datetime import date, datetime
from typing import Any, Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, Field
from app.schemas.common import BaseSchema


class InsurancePolicyBase(BaseSchema):
    company: str = "Insurance Provider"
    provider: Optional[str] = "Insurance Provider"
    policy_number: Optional[str] = None
    policy_holder: Optional[str] = None
    policy_type: str = "Health Insurance"
    plan_name: Optional[str] = None
    policy_name: Optional[str] = None
    coverage_amount: float = Field(default=0.00, ge=0)
    premium_amount: float = Field(default=0.00, ge=0)
    premium: Optional[float] = Field(default=0.00, ge=0)
    premium_frequency: str = "Annual"
    nominee: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    renewal_date: Optional[date] = None
    maturity_date: Optional[date] = None
    claim_contact: Optional[str] = None
    status: str = "Active"
    ocr_confidence: float = 100.0
    document_url: Optional[str] = None
    ai_summary: Optional[str] = None


class InsurancePolicyCreate(InsurancePolicyBase):
    pass


class InsurancePolicyUpdate(BaseModel):
    company: Optional[str] = None
    provider: Optional[str] = None
    policy_number: Optional[str] = None
    policy_holder: Optional[str] = None
    policy_type: Optional[str] = None
    plan_name: Optional[str] = None
    policy_name: Optional[str] = None
    coverage_amount: Optional[float] = Field(default=None, ge=0)
    premium_amount: Optional[float] = Field(default=None, ge=0)
    premium: Optional[float] = Field(default=None, ge=0)
    premium_frequency: Optional[str] = None
    nominee: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    renewal_date: Optional[date] = None
    maturity_date: Optional[date] = None
    claim_contact: Optional[str] = None
    status: Optional[str] = None
    ocr_confidence: Optional[float] = None
    document_url: Optional[str] = None
    ai_summary: Optional[str] = None


class InsurancePolicyResponse(InsurancePolicyBase):
    id: UUID
    user_id: UUID
    created_at: datetime
    updated_at: datetime


class OCRUploadResponse(BaseModel):
    extracted_data: Dict[str, Any]
    analysis: Dict[str, Any]
    ai_explanation: Dict[str, Any]
    validation_warnings: List[str]

File: lib/api/insurance.ts
import { apiClient } from "./client";
import { mockInsurancePolicies } from "@/lib/mock-data";

export interface InsurancePolicyData {
  id?: string;
  policy_name: string;
  policy_number: string;
  insurance_type: string;
  provider_name: string;
  coverage_amount: number;
  premium_amount: number;
  premium_frequency?: string;
  policy_start_date?: string;
  policy_end_date?: string;
  status?: string;
  created_at?: string;
  // UI compatibility aliases
  company?: string;
  provider?: string;
  plan_name?: string;
  policy_type?: string;
  premium?: number;
  analysis?: any;
  ai_explanation?: any;
  ocr_confidence?: number;
  policy_holder?: string;
  nominee?: string;
  claim_contact?: string;
  start_date?: string;
  end_date?: string;
  renewal_date?: string;
  maturity_date?: string;
}

export interface OCRUploadResponseData {
  extracted_data?: Partial<InsurancePolicyData>;
  extracted_info?: Partial<InsurancePolicyData>;
  analysis?: any;
  ai_explanation?: any;
  validation_warnings?: string[];
  raw_text?: string;
  confidence?: number;
}

export async function getInsurancePolicies(): Promise<InsurancePolicyData[]> {
  try {
    const res = await apiClient.get("/insurance");
    const data = res.data?.data?.policies || res.data?.data || res.data;
    if (Array.isArray(data) && data.length > 0) return data;
    return mockInsurancePolicies as InsurancePolicyData[];
  } catch (e) {
    console.warn("Failed to get insurance policies, using mock data:", e);
    return mockInsurancePolicies as InsurancePolicyData[];
  }
}

export async function getPolicies(): Promise<{ policies: InsurancePolicyData[]; portfolio_analysis?: any }> {
  try {
    const res = await apiClient.get("/insurance");
    const data = res.data?.data || res.data;
    if (Array.isArray(data)) {
      return { policies: data };
    }
    if (data?.policies) {
      return data;
    }
    return { policies: mockInsurancePolicies as InsurancePolicyData[] };
  } catch (e) {
    console.warn("Failed to get policies:", e);
    return { policies: mockInsurancePolicies as InsurancePolicyData[] };
  }
}

export async function getPolicy(id: string): Promise<InsurancePolicyData> {
  const res = await apiClient.get(`/insurance/${id}`);
  return res.data?.data || res.data;
}

export async function uploadPolicyDocument(file: File): Promise<OCRUploadResponseData> {
  const formData = new FormData();
  formData.append("file", file);
  const res = await apiClient.post("/insurance/upload", formData, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return res.data?.data || res.data;
}

export async function createPolicy(payload: Partial<InsurancePolicyData>): Promise<InsurancePolicyData> {
  const res = await apiClient.post("/insurance", payload);
  return res.data?.data || res.data;
}

export async function updatePolicy(id: string, payload: Partial<InsurancePolicyData>): Promise<InsurancePolicyData> {
  const res = await apiClient.put(`/insurance/${id}`, payload);
  return res.data?.data || res.data;
}

export async function deletePolicy(id: string): Promise<void> {
  await apiClient.delete(`/insurance/${id}`);
}
